### PR TITLE
Add IF_PRESENT reading mode to materialize

### PR DESCRIPTION
**Pull request by eric-anderson**
Rewrite execution so that we have a finalize() operation that runs if execution runs to completion. This makes reading from materialized output usable because we can now be sure that we have a final execution of the pipeline.

### CODE REVIEW

File: lib/sycamore/sycamore/reader.py [hunk 1]
```diff
@@ -23,6 +24,18 @@ def __init__(self, context: Context, plan: Optional[Node] = None):
         self._context = context
         self.plan = plan
 
+    def materialize(self, path: Union[Path, str], **kwargs) -> DocSet:
```
Comment (HenryL27) [inline lib/sycamore/sycamore/reader.py:27]: can source_mode be an explicit param here?

File: lib/sycamore/sycamore/tests/integration/test_materialize.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from sycamore.context import ExecMode
+import sycamore.tests.unit.test_materialize as test_materialize
+
+
+class TestMaterializeRead(test_materialize.TestMaterializeRead):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.exec_mode = ExecMode.RAY
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/integration/test_materialize.py:8]: is this complete?

File: lib/sycamore/sycamore/tests/unit/test_executor.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import sycamore
+from sycamore.docset import DocSet
+from sycamore.data.document import Document
+from sycamore.plan_nodes import Node
+
+
+def test_finalize():
+    finalize = set()
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/unit/test_executor.py:8]: Does order ever matter?

File: lib/sycamore/sycamore/tests/unit/test_materialize.py [hunk 4]
```diff
@@ -50,13 +51,13 @@ def noop_fn(d):
...
-
-    def tearDown(self):
-        sycamore.shutdown()
+class TestMaterializeWrite(unittest.TestCase):
+    # # Needed until we don't have a global context
+    # def setUp(self):
+    #     sycamore.shutdown()
+    #
+    # def tearDown(self):
+    #     sycamore.shutdown()
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/unit/test_materialize.py:60]: cleaning?

File: lib/sycamore/sycamore/tests/unit/test_materialize.py [hunk 5]
```diff
@@ -50,13 +51,13 @@ def noop_fn(d):
...
 
 
-class TestMaterialize(unittest.TestCase):
-    # Needed until we don't have a global context
-    def setUp(self):
-        sycamore.shutdown()
-
-    def tearDown(self):
-        sycamore.shutdown()
+class TestMaterializeWrite(unittest.TestCase):
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/unit/test_materialize.py:54]: does it need to extend unittest.Testcase?

**Review by HenryL27 — APPROVED** [review]

File: lib/sycamore/sycamore/materialize.py [hunk 6]
```diff
@@ -65,16 +96,56 @@ def ray_callable(ray_input: dict[str, numpy.ndarray]) -> dict[str, numpy.ndarray
 
         return input_dataset
 
+    def _verify_has_files(self) -> None:
+        assert self._root is not None
+        if not self._root.is_dir():
+            raise ValueError(f"Materialize root {self._root} is not a directory")
+
+        for n in self._root.iterdir():
+            if str(n).endswith(".pickle"):
```
[the remark this answers is not in the document]
Comment (eric-anderson) [reply lib/sycamore/sycamore/materialize.py:105]: Done. Thanks.

File: lib/sycamore/sycamore/tests/unit/test_materialize.py (hunk 4 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/tests/unit/test_materialize.py:60]: Done.

**Pull request merged**